### PR TITLE
Clippy book

### PR DESCRIPTION
**Pull request by joshrotenberg**
A work in progress Clippy Book using mdbook. See #6011.

This is currently just a moving around of things:

1. The current README.md split up a bit and put into sections.
1. A rough outline of Clippy lint categories (currently no content, potentially add a basic introduction for each and some example, see questions below.
1. The `docs` content repurposed into a top level `Development` section.
1. The current Roadmap.

Some big questions:

1. is `guide/` the right place? I'm modeling after mdbook itself. 
1. What is the relationship between ALL the Clippy Lints and this guide? It seems like they can coexist. Does that mean the guide should just point to the current side with regard to actual lints, and maybe just include some examples to keep it interesting? Keeping both up to date seems like a maintenance nightmare unless its automated somehow. Or should the current ALL the Clippy lints somehow be incorporated into the book?
1. Related to the above, where should this guide be published since the `gh-pages` branch is already in use?
1. This PR doesn't currently change any existing content. Obviously that would make sense assuming the general structure and relocation is an acceptable approach.

---

Open Tasks for follow up PR:
- Set up CI/CD
- Split up Installation and Usage
- Add more content to Usage (and Installation) chapters
- Enhance CI chapter with more examples for different CIs

---

changelog: The first version of the *Clippy Book*
staring, *The Clippy Team*, *Bors* and *rustc*

### CODE REVIEW

File: guide/src/SUMMARY.md
Comment (flip1995) [inline guide/src/SUMMARY.md:23]: It may make sense to put these in their own category names "Infrastructure", since normal contributors usually don't have to interact with those.

File: guide/src/SUMMARY.md
Comment (flip1995) [inline guide/src/SUMMARY.md:25]: I hope we'll do more roadmaps in the future. So I would put this in it's own category:

```suggestion
- [Roadmap](...)
    - [2021](roadmap-2021.md)
```

File: guide/src/configuration.md
Comment (flip1995) [inline guide/src/configuration.md:1]: This should be at the end of the configuration doc. Also I'd like to rewrite this section a bit (noting that this config file was never stabilized).

**Review by flip1995 — COMMENTED** [review]
Thanks! This looks already way better than the previous attempt. This is a good starting point in working out how the book should look like cc @rust-lang/clippy 

Some things that should be done in this PR (not necessarily by you):
- There should also exist documentation on how to build the book (you can put this in "Development" or "Infrastructure" section)
- Our automation probably have to be adapted to this (not sure what exactly yet, but I'd assume `clippy_dev` and the metadata collection, depending on what we want to do with this)

---

>     2. A rough outline of Clippy lint categories

I like this. I even would like to put an automation in place that puts all lints from the specific category in those files, linking to the documentation with a short description of the lint. (We should first switch to the new metadata collection though #7298)

---

>     1. is `guide/` the right place? I'm modeling after mdbook itself.

Why not `book/`?


>     2. What is the relationship between ALL the Clippy Lints and this guide?

TBH we're not quite sure yet. For now they should coexist. For your other questions, see above.

>     3. Related to the above, where should this guide be published since the `gh-pages` branch is already in use?

Good question. Probably also on the gh-pages branch, but add a link on the website (in addition to the current `README`) to the book. We can just push the artifacts to a `book/` dir on the `gh-pages` branch.

>     4. This PR doesn't currently change any existing content.

That's fine. But we should rewrite the README and remove the duplicated documentation int `doc/`. The README should probably only contain a link to the book and a quick-start guide to Clippy. That quick-start guide should probably mention lint groups, how to allow/deny lints and link to the book for things like how to configure Clippy.

File: guide/src/SUMMARY.md
Comment (joshrotenberg) [reply guide/src/SUMMARY.md]: done

File: guide/src/SUMMARY.md
Comment (joshrotenberg) [reply guide/src/SUMMARY.md]: I added a roadmap section for these.

**Comment by flip1995** [comment]
@joshrotenberg I want to first get our infra move through, before I can focus on this PR, since this kind of depends on it¹. So expect that it will take about 2 weeks before we can really make progress with this PR. I hope that's ok with you. 

Also, since I expect that this PR will include many incremental changes, could you avoid amending the one commit in this PR and just add a new commit to change something? Also please only force push if you have to do a rebase. That will make reviewing and working on this way easier.

---
¹: I want to get some automation in it in this PR already and we would do the work twice if we'd do it with the current infra and then again for the new one.

File: guide/src/configuration.md
Comment (joshrotenberg) [reply guide/src/configuration.md]: > This should be at the end of the configuration doc. 

Not exactly sure what you mean here. 

> Also I'd like to rewrite this section a bit (noting that this config file was never stabilized).

Yeah, I imagine a lot of stuff can use some cleanup and rewriting. I'm happy to help with that as well.

**Comment by flip1995** [comment]
Good news: The metadata collector move is done, so getting this through will be my next project. :rocket:

**Comment by joshrotenberg** [comment]
> Good news: The metadata collector move is done, so getting this through will be my next project. 🚀

Great! I'd love to get back to this!

I made a few minor updates just the other day. Going back through some of the comments above:

I added a short blurb about how to do this. I'm sure it can be improved.

I think your idea about just publishing to `gh-pages` into a `book` directory and linking from the index of what is currently there makes sense. I took a brief look at the `deploy.yml`: maybe it's just another `step` in there?

There is also the question of the `Lints` section in the book: a) make it static content with more of an overview (with links to the Clippy Lints site), b) do something dynamic there like you mentioned, c) leave it out for now since Clippy Lints has everything someone might need anyway.

And then removing duplicate content from the `README.md` and the `doc` directory: remove it in this PR? Get everything else squared away and do another PR to migrate?

I'm happy to work on, collaborate and/or relinquish any of the above, just let me know your thoughts.

**Comment by flip1995** [comment]
> I think your idea about just publishing to `gh-pages` into a `book` directory and linking from the index of what is currently there makes sense.

I wouldn't include this in this PR, but take care of this once the book is merged. This was also a successful approach with the metadata collector. First implement everything and then switch the infra to that.

> c) leave it out for now

I would leave it out for this PR, but I'd like a list of the lints in the book later. The primary documentation should live on the website though. So a static page linking to the website should be good for now.



> And then removing duplicate content from the `README.md` and the `doc` directory

This should be a quick thing to do, so I would do this once we move everything over to the book.

File: book/src/configuration.md
Comment (flip1995) [inline book/src/configuration.md:18]: This may also fit in the `Usage` chapter.

File: book/src/development/README.md
Comment (flip1995) [inline book/src/development/README.md:1]: I think we can add some text here from the CONTRIBUTING.md, but I'm not sure what exactly, yet. Same for Infrastructure.

File: book/src/development/basics.md
Comment (flip1995) [inline book/src/development/basics.md:1]: I didn't review the sub-chapters of the development chapter, but right before merging this, we should re-copy those files, so those are up-to-date.  Same for Infrastructure.

File: book/src/roadmap/2021.md
Comment (flip1995) [inline book/src/roadmap/2021.md:1]: The roadmap/README.md file is currently empty, and we should also add some text there. I'll still have to think about something though.

**Review by flip1995 — COMMENTED** [review]
First round of review with suggestions to better structure this book. Let me know if you disagree with anything!

File: book/src/README.md
Comment (joshrotenberg) [reply book/src/README.md]: Updated.

File: book/src/installation_and_usage.md
Comment (joshrotenberg) [reply book/src/installation_and_usage.md:4]: I added a chapter for CI and linked to it from here. Added a page for GitHub Actions and a placeholder for Gitlab for now.

File: book/src/installation_and_usage.md
Comment (joshrotenberg) [reply book/src/installation_and_usage.md:8]: Added a quick blurb here, open to suggestions.

File: book/src/installation_and_usage.md
Comment (joshrotenberg) [reply book/src/installation_and_usage.md:67]: I think splitting them up makes sense just to make the steps clearer, i.e. "I know clippy is installed (since I'm already using rust), tell me how to use it".

I can split the two up, and then we can iterate on the actual content.

File: book/src/configuration.md
Comment (joshrotenberg) [reply book/src/configuration.md:1]: Added this.

File: book/src/installation_and_usage.md
Comment (joshrotenberg) [reply book/src/installation_and_usage.md:67]: Actually I think I've changed my mind on this for now. It's currently pretty short and still covers the basics. I suggest we leave it as is, and then consider a more in depth standalone usage page in the next round with much more detail, including everything currently there plus more in depth command line and  in code usages.

File: book/src/continuous_integration/github_actions.md
Comment (flip1995) [inline book/src/continuous_integration/github_actions.md:12]: NIT: Usually your indentation for those files is 2 spaces.

File: book/src/continuous_integration/github_actions.md
Comment (flip1995) [inline book/src/continuous_integration/github_actions.md:20]: This is not necessary. The latest stable Rust toolchain is already installed including Clippy on all GitHub provided runners. So just 
```suggestion
            -   name: Run Clippy
                run: cargo clippy
```

already does the job.

Comment (flip1995) [reply book/src/continuous_integration/github_actions.md:20]: For the different flags, I would just copy it like it is documented in the Travis section.

File: .github/workflows/remark.yml
Comment (flip1995) [inline .github/workflows/remark.yml:24]: Rebasing onto master should include this.

File: book/src/development/basics.md
Comment (flip1995) [inline book/src/development/basics.md:169]: Same here. Rebase should fix this.

File: book/src/installation_and_usage.md
Comment (flip1995) [inline book/src/installation_and_usage.md:35]: Since Clippy is most likely pre-installed I would split up installation and usage. Most people will only be interested in the usage part.

---

EDIT:

> Actually I think I've changed my mind on this for now. It's currently pretty short and still covers the basics. I suggest we leave it as is, and then consider a more in depth standalone usage page in the next round with much more detail, including everything currently there plus more in depth command line and in code usages.

Just saw your comment here. This makes sense, we can make the split in a follow up PR. ~~Could you track open tasks (not addressed in this PR) in the PR body please?~~ I put it into the PR body.

File: book/src/configuration.md
Comment (flip1995) [inline book/src/configuration.md:3]: You can crate blocks, by writing this as

```suggestion
> **Note:** The configuration file is unstable and may be deprecated in the future.
```

At least this is what the Rust reference does: https://github.com/rust-lang/reference/blob/master/src/introduction.md#conventions

File: book/src/continuous_integration/travis.md
Comment (flip1995) [inline book/src/continuous_integration/travis.md:22]: Probably worth copying this to all configurations? Or for DRY purposes, this should probably be moved to the usage part?

**Review by flip1995 — COMMENTED** [review]
Next round of review.

File: book/src/continuous_integration/travis.md
Comment (joshrotenberg) [reply book/src/continuous_integration/travis.md:22]: Agree with moving it to usage, and then maybe in the top level of `continuous_integration.md` we say something like "check the usage docs for all of the command line options".

File: book/book.toml
Comment (flip1995) [inline book/book.toml:2]: Is the authors field necessary? If so, please change this to 
```suggestion
authors = ["The Rust Clippy Developers"]
```

If not, just remove it.

File: book/book.toml
Comment (flip1995) [inline book/book.toml:13]: I think those URLs are outdated?

File: book/book.toml
Comment (flip1995) [inline book/book.toml:14]: Do we need this for the Clippy documentation?

**Review by flip1995 — COMMENTED** [review]
A few more comments. Please also fix this: https://github.com/rust-lang/rust-clippy/pull/7359#discussion_r682438639

After that, I think I'll write some text myself for the currently empty chapters and then this should be good to go.

File: book/book.toml
Comment (joshrotenberg) [reply book/book.toml:2]: Worth having in there I think.

File: book/src/continuous_integration/github_actions.md
Comment (joshrotenberg) [reply book/src/continuous_integration/github_actions.md:12]: Sorry, missed this one. Fixed.

File: book/book.toml
Comment (joshrotenberg) [reply book/book.toml:14]: Only if we want the GitHub icon and a link to edit the docs (for example, on the mdbook guide it links to this: https://github.com/rust-lang/mdBook/tree/master/guide). We can just leave it out, too.

File: book/book.toml
Comment (joshrotenberg) [reply book/book.toml]: See below. I fixed them, but we can also remove them.

File: book/book.toml
Comment (flip1995) [reply book/book.toml:14]: It won't hurt to leave them in I assume. So nothing to do here.

Comment (joshrotenberg) [reply book/book.toml:14]: I changed it to `book` from `guide` so when we do actually deploy it it should point to the right place.

**Comment by flip1995** [comment]
I'll summarize all the open points that are left in this PR, to have it in one place instead of spread out over multiple review comment threads. I'll still have to figure out what to address in this PR and what can be addressed later.

- [ ] Clean up / split up `CONTRIBUTING.md` and move the contents into the book (possibly new sections/chapters)
- [x] Slim down the `README.md` and just link to the book.
- [ ] Add text to chapter introductions (Possibly from `CONTRIBUTING.md`)
    - [ ] Clippy Development
    - [ ] Infrastructure
    - [ ] Roadmap
    - [ ] Clippy Lints (maybe remove the draft sections for now)
    - [ ] CI (Also general cleanup in the whole chapter)
- [ ] Figure out what to change in `clippy_dev`
- [ ] Move `doc/` contents completely to the book instead of copying them
- [x] Split up Installation and Usage (those probably will need a rewrite)
- [x] Move note from "Travis CI" section to usage and link to it from each CI section
- [x] Enhance CI chapter with more examples for different CIs
- [x] Set up CI/CD for deployment of the book

I checked off the once I think can/should be done in follow up PRs

---

I want to get this book into the repo in a "first release" state instead of a "first draft" state. That lessens the the pressure of directly opening and merging follow up PRs once this gets merged.

I'll address the above points myself, so nothing to do for you @joshrotenberg for now. :+1:

**Comment by joshrotenberg** [comment]
Sorry, I've been on vacation for a bit. I'm back now, feel free to throw anything my way on this if you'd like.

**Comment by flip1995** [comment]
It's mostly stuff I have to take a look at myself first. But once I got to it and find things, I can throw your way, I'll definitely do so!

**Comment by joshrotenberg** [comment]
Just checking in to see if I can help out here.

**Comment by flip1995** [comment]
I didn't had time to get to it. I started working on it, but just for a few hours, so there's still a bit to do for me, before this can move forward.

I plan to address this over the holidays and now interpret this as my Christmas present for Clippy :)

**Comment by flip1995** [comment]
I didn't manage to work on this over the holidays. I was too busy enjoying my vacation. I have some more time I can spend on Clippy this month, so I try to get this out of the backlog.

**Comment by joshrotenberg** [comment]
> I didn't manage to work on this over the holidays. I was too busy enjoying my vacation. I have some more time I can spend on Clippy this month, so I try to get this out of the backlog.

Hope you had a great vacation! As always, let me know if I can help out here.

**Comment by flip1995** [comment]
Alright, I did some work (still not finished though). I also updated my todo comment from above accordingly.

I rebased this branch on the latest master (clean rebase, no changes of Josh's commits).

**Comment by flip1995** [comment]
Almost ready for a final review. The only things left to do are:

- Write short paragraph for "Infrastructure" chapter.
- Small update to `cargo dev update_lints`, so that the lint counter in the book also gets updated.
- Figure out why remark fails

**Comment by flip1995** [comment]
@rustbot ready :rocket: 

@joshrotenberg I finally got to finish the first version of the book and got it to a point, where I think that it is ready to be released. Thanks so much for the initial work you put in and sorry that it took me so long to give it the final touches.

Things left TODO in future PRs:

> - [ ] Enhance CI chapter with more examples for different CIs 
> - [ ] Set up CI/CD for deployment of the book
> - [ ] Slim down the `README.md` and just link to the book.

**Comment by flip1995** [comment]
I also have a commit ready to deploy the book on CI: https://github.com/flip1995/rust-clippy/commit/e40c29756ddec1579c4448a484d6c843e824ef5e But I will open a separate PR for it once this is merged.

**Comment by xFrednet** [comment]
Congratulations, really happy to see this project coming to an end!!! :+1:

File: book/src/usage.md
Comment (xFrednet) [inline book/src/usage.md:88]: I would move this section to or combine it with the *Error on lints* section. We can maybe also include a link to [rustc's book > *compiler flags*](https://doc.rust-lang.org/rustc/lints/levels.html#via-compiler-flag) for further information.

File: book/src/lints.md
Comment (xFrednet) [inline book/src/lints.md:28]: > _sus_

Oh no no no no no no no, I like it :joy:

File: book/src/lints.md
Comment (xFrednet) [inline book/src/lints.md:34]: Maybe?:

```suggestion
that offends the lint you should specifically and locally `#[allow]` the lint and add give a
```

**Review by xFrednet — COMMENTED** [review]
Awesome work, from the parts I've read so far! I've added some comments, a bit nit-picky as usual, so feel free to not apply every suggestion.

`3/26` files, that makes about 12 %

File: book/src/usage.md
Comment (flip1995) [reply book/src/usage.md:44]: I think we use a feature for this to also fail path dependencies, i.e. `clippy_lints` and `clippy_utils`. `-D warnings` isn't propagate through to those AFAIK.

**Comment by flip1995** [comment]
Addressed all comments. I will hold off with rebasing until everything is reviewed. All the files that are moved from `doc/` are unchanged, except for the formatting. So you don't really have to review those.

File: book/src/development/README.md
Comment (xFrednet) [inline book/src/development/README.md:38]: Should we maybe also link to rustc's dev guide for these topics?

File: book/src/development/adding_lints.md
Comment (xFrednet) [inline book/src/development/adding_lints.md:120]: At some point we should also add some documentation for method lints and the methods module. Just a reminder, having is book is already a big step and we can probably create an issue for someone to write additional documentation 🙃

**Review by xFrednet — COMMENTED** [review]
Some small nits but nothing major. I guess I've read about 50% of the files now :)

File: book/src/development/README.md
Comment (flip1995) [reply book/src/development/README.md:38]: I would rather do that in the `CONTRIBUTING.md` file (which probably already does that) and later once we move those things to the book in those files. This is just a TODO list for the book itself.

File: book/src/development/adding_lints.md
Comment (flip1995) [reply book/src/development/adding_lints.md:120]: Agreed, I don't want to add too many new things in the PR though.

**Review by xFrednet — COMMENTED** [review]
Some things I found while trying to do a dummy sync :upside_down_face:

File: book/src/development/infrastructure/release.md
Comment (xFrednet) [inline book/src/development/infrastructure/release.md:21]: For the scripts in this section, it would be good to know in which repo they should be executed

File: book/src/development/infrastructure/release.md
Comment (xFrednet) [inline book/src/development/infrastructure/release.md:50]: There is a section where we check something in the beta branch of rust. I think it would be better to state:

```
git fetch upstream
git checkout upstream/beta
BETA_SHA=$(git log --oneline -- src/tools/clippy/ | grep -o "Merge commit '[a-f0-9]*' into .*" | head -1 | sed -e "s/Merge commit '\([a-f0-9]*\)' into .*/\1/g")
```

**Review by xFrednet — COMMENTED** [review]
Two small comments regarding the release docs. The process was pretty strait forward :)

I'm curious, have you created some scripts/aliases to automate the process a bit? :upside_down_face:

**Review by xFrednet — COMMENTED** [review]
image

:tada:

I've looked through every file. I didn't really check spelling, since that's not my specialty ^^. The content looks good. As you see, I've added comments where I think the documentation could be improved :upside_down_face:

File: book/src/installation.md
Comment (flip1995) [reply book/src/installation.md:20]: I'm open to moving the install from source section to here. However this is user documentation not dev documentation, so I don't see the need of mentioning `cargo dev` at all.

For the install from source section: I intentionally left this in the developer guide to signal that this is not a recommended way of installing Clippy.

File: book/src/development/proposals/README.md
Comment (flip1995) [reply book/src/development/proposals/README.md:3]: This is for accepted proposals.

File: book/src/development/infrastructure/release.md
Comment (flip1995) [reply book/src/development/infrastructure/release.md:21]: I think I always mention this. But it seems to not be clear enough. I try to improve on that.

File: book/src/development/infrastructure/sync.md
Comment (flip1995) [reply book/src/development/infrastructure/sync.md:81]: I also hate those merge commits, because it destroys the `git blame` history and you aren't able to click through this in the GitHub UI anymore. But I don't think there is anything we can do about that...

File: book/src/development/infrastructure/sync.md
Comment (flip1995) [reply book/src/development/infrastructure/sync.md:72]: This is old documentation that I wrote before I knew how to do this properly. I'll improve that.

File: book/src/development/infrastructure/release.md
Comment (xFrednet) [reply book/src/development/infrastructure/release.md:21]: This was the only case, where it wasn't quite clear :)

File: book/src/installation.md
Comment (xFrednet) [reply book/src/installation.md:20]: > I don't see the need of mentioning `cargo dev` at all.

Since, I've started contributing to Clippy, we had two or three people who asked about installing from source. One of them wanted to have Clippy for a platform that required the installation from source, while the others intended to check their changes on a local crate. The reference of `cargo dev` is intended for the second user group. I'm fine if you leave it out if you don't agree. :upside_down_face:

Comment (flip1995) [reply book/src/installation.md:20]: I count changing something in Clippy and then wanting to test it as Clippy dev, so they should look into the Clippy dev documentation. 

I really don't want to have dev tools and non-recommended installation methods in the user documentation, because that makes the impression, that we actively support that. And since we don't test those things in CI, I don't think we can safely say that we actively support it (for the general user).

**Comment by flip1995** [comment]
Things left TODO in future PRs:

- [ ] Enhance CI chapter with more examples for different CIs
- [ ] Set up CI/CD for deployment of the book
- [ ] Slim down the `README.md` and just link to the book.
- [ ] Better explain lints in submodules, like `methods`/`matches`/... https://github.com/rust-lang/rust-clippy/pull/7359#discussion_r872404196
- [ ] Publish Clippy book through doc.rs/doc.rust-lang.org: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/doc.2Erust-lang.2Eorg.20URL.20for.20Clippy/near/279424472

**Review by xFrednet — APPROVED** [review]
The latest version looks good to me, including the TODO list posted in this PR.

@flip1995 If you want to merge this, you can r=me, after a rebase; otherwise, I'm also happy to review further adjustments. :upside_down_face:

**Comment by flip1995** [comment]
I squashed a bunch of commits and rebased on the latest master without any changes to the last state you just reviewed. (except for the conflict resolution in adding_lints and common_tools of course).

**Comment by xFrednet** [comment]
The number of commits looks reasonable to me and there are no filler ones, so it's fine. I guess this marks the beginning of an era ^^

A big thank you to @joshrotenberg for taking the initiative and starting to write the book. And thank you, @flip1995, for reviewing the first part, adding new chapters and applying my suggestions.

Without further of do, I'll hand this off to our merging wizard bors:

@bors r+

**Comment by xFrednet** [comment]
The changelog entry was missing, I've added one

@bors retry

**Pull request merged**
